robot-server/tests/integration/http_api/persistence/test_compatibility.py

```diff
@@ -49,39 +68,55 @@ async def test_protocols_analyses_and_runs_available_from_older_persistence_dir(
-                for analysis_id in analysis_ids:
-                    await robot_client.get_analysis(
-                        protocol_id=protocol_id, analysis_id=analysis_id
-                    )
-
-            all_runs = (await robot_client.get_runs()).json()
-            all_run_ids = [r["id"] for r in all_runs["data"]]
-
-            assert len(all_run_ids) == snapshot.expected_run_count
-
-            for run_id in all_run_ids:
-                await robot_client.get_run(run_id=run_id)
-
-                all_command_summaries = (
-                    await robot_client.get_run_commands(
-                        run_id=run_id,
-                        page_length=999999,  # Big enough to include all commands.
-                    )
-                ).json()
-                assert len(all_command_summaries["data"]) > 0
-
+            all_protocols = (await robot_client.get_protocols()).json()["data"]
+
+            assert len(all_protocols) == snapshot.expected_protocol_count
+
+            for protocol_from_all_protocols_endpoint in all_protocols:
+                protocol_id = protocol_from_all_protocols_endpoint["id"]
+
+                analyses_from_all_analyses_endpoint = (
+                    await robot_client.get_analyses(protocol_id=protocol_id)
+                ).json()["data"]
+
+                analysis_ids_from_all_protocols_endpoint = [
+                    a["id"]
+                    for a in protocol_from_all_protocols_endpoint["analysisSummaries"]
+                ]
+
+                analysis_ids_from_all_analyses_endpoint = [
+                    a["id"] for a in analyses_from_all_analyses_endpoint
+                ]
+
+                assert (
+                    analysis_ids_from_all_protocols_endpoint
+                    == analysis_ids_from_all_analyses_endpoint
+                )
+
+                number_of_analyses = len(analysis_ids_from_all_protocols_endpoint)
+                if protocol_id in snapshot.protocols_with_no_analyses:
+                    assert number_of_analyses == 0
+                else:
+                    assert number_of_analyses > 0
+
+                all_runs = (await robot_client.get_runs()).json()
+                all_run_ids = [r["id"] for r in all_runs["data"]]
+                assert len(all_run_ids) == snapshot.expected_run_count
+
+                for run_id in all_run_ids:
+                    await robot_client.get_run(run_id=run_id)
+
+                    all_command_summaries = (
+                        await robot_client.get_run_commands(
+                            run_id=run_id,
+                            page_length=999999,  # Big enough to include all commands.
+                        )
+                    ).json()
+
+                if len(all_command_summaries["data"]) == 0:
+                    assert run_id in snapshot.runs_with_no_commands
+                else:
+                    assert len(all_command_summaries["data"]) > 0
```

Comment:
Is this block mis-indented? It looks like it should be inside the `for run_id in all_run_ids:` loop.

Also, I would rewrite the check like this:

```suggestion
                    if run_id in snapshot.runs_with_no_commands:
                        assert len(all_command_summaries["data"]) == 0
                    else:
                        assert len(all_command_summaries["data"]) > 0
```

For symmetry with the `number_of_analyses` check.